fairmq/sdk/Topology.h

```diff
@@ -81,82 +86,325 @@ inline bool StateEqualsTo(const TopologyState& topologyState, DeviceState state)
+    ///           case fair::mq::ErrorCode::DeviceChangeStateFailed:
+    ///             // failed to change state of a fairmq device
+    ///           case fair::mq::ErrorCode::OperationInProgress:
+    ///             // async operation already in progress
+    ///           default:
+    ///         }
+    ///     }
+    /// }
+    /// @endcode
+    /// With coroutine (C++20, see https://en.cppreference.com/w/cpp/language/coroutines):
+    /// @code
+    /// try {
+    ///     fair::mq::sdk::TopologyState state = co_await
+    ///         topo.AsyncChangeState(fair::mq::sdk::TopologyTransition::InitDevice,
+    ///                               std::chrono::milliseconds(500),
+    ///                               asio::use_awaitable);
+    ///     // success
+    /// } catch (const std::system_error& ex) {
+    ///     if (ex.code().category().name() == "fairmq") {
+    ///         switch (static_cast<fair::mq::ErrorCode>(ec.value())) {
+    ///           case fair::mq::ErrorCode::OperationTimeout:
+    ///             // async operation timed out
+    ///           case fair::mq::ErrorCode::OperationCanceled:
+    ///             // async operation canceled
+    ///           case fair::mq::ErrorCode::DeviceChangeStateFailed:
+    ///             // failed to change state of a fairmq device
+    ///           case fair::mq::ErrorCode::OperationInProgress:
+    ///             // async operation already in progress
+    ///           default:
+    ///         }
+    ///     }
+    /// }
+    /// @endcode
+    template<typename CompletionToken>
+    auto AsyncChangeState(TopologyTransition transition,
+                          Duration timeout,
+                          CompletionToken&& token)
+    {
+        return asio::async_initiate<CompletionToken, ChangeStateCompletionSignature>(
+            [&](auto handler) {
+                std::lock_guard<std::mutex> lk(fMtx);
+
+                if (fChangeStateOp.IsCompleted()) {
+                    fChangeStateOp = ChangeStateOp(AsioBase<Executor, Allocator>::GetExecutor(),
+                                                   AsioBase<Executor, Allocator>::GetAllocator(),
+                                                   std::move(handler));
+                    fChangeStateTarget = expectedState.at(transition);
+                    fDDSSession.SendCommand(GetTransitionName(transition));
+                    if (timeout > std::chrono::milliseconds(0)) {
+                        fChangeStateOpTimer.expires_after(timeout);
+                        fChangeStateOpTimer.async_wait([&](std::error_code ec) {
+                            if (!ec) {
+                                std::lock_guard<std::mutex> lk2(fMtx);
+                                fChangeStateOp.Timeout(fState);
+                            }
+                        });
+                    }
+                } else {
+                    throw std::system_error(MakeErrorCode(ErrorCode::OperationInProgress),
```

Comment:
@rbx Just realized this should no longer be an exception but an immediate completion with the same error code. Missed that one before.

Reply:
Do you want to change this in this PR?

Reply:
yes

Reply:
fixed now